dfd-wrapper/src/main/java/com/revolut/rxdata/dfd_wrapper/DataFlowDelegate.kt

```diff
@@ -0,0 +1,109 @@
+package com.revolut.rxdata.dfd_wrapper
+
+import com.revolut.data.model.Data
+import com.revolut.rxdata.dod.DataObservableDelegate
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.rx2.asFlow
+import kotlinx.coroutines.rx2.await
+import kotlinx.coroutines.rx2.rxSingle
+
+/*
+ * Copyright (C) 2023 Revolut
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+class DataFlowDelegate<Params : Any, Domain : Any>(
+    fromNetwork: suspend (params: Params) -> Domain,
+    fromMemory: (params: Params) -> Domain?,
+    toMemory: (params: Params, Domain) -> Unit,
+    fromStorage: suspend (params: Params) -> Domain?,
+    toStorage: suspend (params: Params, Domain) -> Unit,
+    onRemove: suspend (params: Params) -> Unit = { _ -> Unit },
+) {
+    private val inner = DataObservableDelegate<Params, Domain>(
+        fromNetwork = { params ->
+            rxSingle(DataFlowDelegateDispatchers.ioDispatcher()) {
+                fromNetwork(params)
+            }
+        },
+        fromMemory = { params ->
+            fromMemory(params)
+        },
+        toMemory = { params, domain ->
+            toMemory(params, domain)
+        },
+        fromStorage = { params ->
+            runBlocking {
+                fromStorage(params)
+            }
+        },
+        toStorage = { params, domain ->
+            runBlocking {
```

Comment:
Looks dangerous. When client calls `updateStorage`, we will block the thread

Reply:
But this is exactly the same in DOD. We can make `updateStorage` & `updateAll` suspendable and switch dispatcher there. I assume we must first save and only then notify observers, don't we?

Reply:
You are right. Seems like a bug in DOD to me. I would vote to make `updateStorage` and `updateAll` suspendable and switch dispatcher to `IO` there. And I think you are right, and it is better to notify observers after saving

Reply:
Done. 😁

Reply:
sorry I'm confused, shouldn't we remove runBlocking {} from lambdas now?

Reply:
The problem is DOD's `fromStorage` lambda is synchronous, while ours is suspending (to simplify usage for cases when it should call some existing suspending API). But it's still safe to call `update*` methods from any thread, since we switch to IO dispatcher first and _then_ block current thread.

If we remove `runBlocking`, we have two choices:
* make our `fromStorage` synchronous, users will have to wrap suspending APIs themselves if needed
* use `someCoroutineScope.launch { fromStorage(params) }`, but then we lose consistency guarantees (it can execute, and possibly fail, after observers are notified) and need to deal with errors somehow.

Reply:
@yatsinar So should we keep it as-is?

Reply:
Gotcha, let's keep runBlocking then